### PR TITLE
[v4] feat: migrate to dart-sass for sass compilation

### LINKED ISSUE #4032: Compilation issue with newer version of SASS
When trying to import to blueprint using SASS, error is encountered.

#### Environment
- __Package version(s)__:
  - @blueprintjs/core: 3.24.0
  - webpack: 4.42.1
  - sass: 1.26.3
  - sass-loader: 8.0.2
- __Browser and OS versions__:  MacOS Catalina 10.15.4

#### Steps to reproduce

1. Create a SASS file and add `@import "~@blueprintjs/core/src/blueprint";`
2. Try compiling it using webpack.

#### Actual behavior

```
ModuleBuildError: Module build failed (from ./node_modules/sass-loader/dist/cjs.js):
SassError: compound selectors may no longer be extended.
Consider `@extend .bp3-menu-item, :hover` instead.
See http://bit.ly/ExtendCompound for details.

   ╷
19 │       @extend .#{$ns}-menu-item:hover;
   │               ^^^^^^^^^^^^^^^^^^^^^^^
   ╵
  node_modules/@blueprintjs/core/src/components/menu/_submenu.scss 19:15  root stylesheet
```

#### Expected behavior

There should be no error.

### PR DESCRIPTION
**Pull request by ianpf**
#### Checklist

- [ ] Includes tests (N/A?)
- [X] Update documentation

#### Changes proposed in this pull request:

Migrating from node-sass to the (now) reference implementation, dart-sass.

This is _not yet ready to merge_. I know that I need to fix the inline-svg usage, and replace it with an alternative. If anyone has any suggestions here I'd love input.

I'm not totally clear on the removal of the `~`. I'm sure I'm just missing something in dart-sass which makes it work, and again very happy for any input.

#### Reviewers should focus on:

Please note, that `packages/stylelint-plugin/test/no-prefix-literal.test.js` fails on `develop`. I'm guessing that `bp-` use to be reserved alongside `bp3`/`bp4`.

#### Screenshot
N/A

### CODE REVIEW

**Comment by palantirtech** [comment]
Thanks for your interest in [palantir/blueprint](https://github.com/palantir/blueprint), @ianpf! Before we can accept your pull request, you need to sign our contributor license agreement - just visit https://cla.palantir.com/ and follow the instructions. Once you sign, I'll automatically update this pull request.

**Comment by ianpf** [comment]
I think that I can probably use the old svg-inline package, if I change to used the JavaScript API for the `Sass` package. It looks like it supports node-sass like plugins here: https://sass-lang.com/documentation/js-api#functions

File: packages/datetime/src/blueprint-datetime.scss [hunk 1]
```diff
@@ -1,13 +0,0 @@
-/*
```
Comment (adidahiya) [inline packages/datetime/src/blueprint-datetime.scss]: why was this file deleted?

File: packages/node-build-scripts/sass-compile.js [hunk 2]
```diff
@@ -0,0 +1,111 @@
+#!/usr/bin/env node
+/** Copyright 2020 Palantir Technologies, Inc. All rights reserved.
```
Comment (adidahiya) [inline packages/node-build-scripts/sass-compile.js]:
```suggestion
/**
 * Copyright 2021 Palantir Technologies, Inc. All rights reserved.
```

File: packages/node-build-scripts/sass-compile.js [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+
+const argv = yargs.parse(hideBin(process.argv));
+
+let functions;
+if (argv.functions) {
+    const functionsPath = path.join(process.cwd(), argv.functions);
+    try {
+        functions = require(functionsPath);
+    } catch (ex) {
+        console.log(`Could not find ${functionsPath}`);
```
Comment (adidahiya) [inline packages/node-build-scripts/sass-compile.js]:
```suggestion
        console.error(`[node-build-scripts] Could not find ${functionsPath}`);
```

File: packages/node-build-scripts/sass-compile.js [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+        clearTimeout(timeout);
+        timeout = setTimeout(() => func(...args), time);
+        return () => clearTimeout(timeout);
+    };
+}
+
+function compileDirectory({ input, functions, output }) {
+    const files = fs.readdirSync(input).map(value => path.join(input, value));
+    files
+        .filter(value => !path.basename(value).startsWith("_") && value.match(/\.s[ac]ss$/))
```
Comment (adidahiya) [inline packages/node-build-scripts/sass-compile.js]: it's a little unfortunate that we have to do this filtering manually. seems like `node-sass`'s CLI was handling this feature of the language for us. let's at least leave a code comment. also I suggest flipping the expression order for better readability.

```suggestion
        // filter out Sass partials (filenames start with underscore)
        .filter(value => value.match(/\.s[ac]ss$/) && !path.basename(value).startsWith("_"))
```

File: packages/node-build-scripts/sass-compile.js [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+function compilePath({ input, functions, output, watch }) {
+    output = output || input.replace(/\.s[ac]ss$/, ".css");
+    const inputStat = fs.statSync(input);
+    if (!inputStat.isDirectory()) {
+        const result = compileFile({ input, functions, output });
+        if (watch) {
+            const watcher = chokidar.watch(result.stats.includedFiles);
+            watcher.once("ready", () => {
+                watcher.on("all", (event, payload) => {
+                    const result = compileFile({ input, functions, output });
```
Comment (adidahiya) [inline packages/node-build-scripts/sass-compile.js]: does this return value need to be stored as a variable? it's not being used after compilation

File: packages/node-build-scripts/sass-compile.js [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+                });
+            });
+        }
+    } else {
+        if (watch) {
+            const watcher = chokidar.watch(input);
+            console.log(`watching ${input}`);
```
Comment (adidahiya) [inline packages/node-build-scripts/sass-compile.js]:
```suggestion
            console.info(`[node-build-scripts] Watching ${input}`);
```

File: packages/stylelint-plugin/src/utils/cssSyntax.ts [hunk 7]
```diff
@@ -34,8 +34,8 @@ export const BpPrefixVariableMap: Record<Exclude<CssSyntax, CssSyntax.OTHER>, st
 };
 
 export const BpVariableImportMap: Record<Exclude<CssSyntax, CssSyntax.OTHER>, string> = {
-    [CssSyntax.SASS]: "~@blueprintjs/core/lib/scss/variables",
-    [CssSyntax.LESS]: "~@blueprintjs/core/lib/less/variables",
+    [CssSyntax.SASS]: "@blueprintjs/core/lib/scss/variables",
```
Comment (adidahiya) [inline packages/stylelint-plugin/src/utils/cssSyntax.ts:37]: hmm, this part will get a bit tricky because we have a lot of other tooling which still uses node-sass. we'll need to make this import syntax configurable at the lint rule or plugin level somehow.

File: packages/webpack-build-scripts/package.json [hunk 8]
```diff
@@ -15,7 +15,6 @@
         "fork-ts-checker-notifier-webpack-plugin": "^3.0.0",
         "fork-ts-checker-webpack-plugin": "^6.1.0",
         "mini-css-extract-plugin": "^1.3.5",
-        "node-sass": "^5.0.0",
```
Comment (adidahiya) [inline packages/webpack-build-scripts/package.json:18]: I believe we need to add a `"sass"` dependency here after this change: https://webpack.js.org/loaders/sass-loader/#implementation

File: packages/datetime/src/blueprint-datetime.scss (hunk 1 above)
Comment (adidahiya) [reply packages/datetime/src/blueprint-datetime.scss]: the removal of this file is causing sass compilation in the datetime package to fail since there are no non-partial Sass files to compile. you'll need to put it back.

from the build log:

```
@blueprintjs/datetime: Input Error: You must pass a valid list of files to parse
```

**Review by adidahiya — CHANGES_REQUESTED** [review]
overall, some great progress here!

File: packages/stylelint-plugin/src/utils/cssSyntax.ts (hunk 7 above)
Comment (adidahiya) [reply packages/stylelint-plugin/src/utils/cssSyntax.ts:37]: actually, it looks like we already have this configurable for both stylelint rules:

- https://github.com/palantir/blueprint/blob/67c9e2ec477f0c189782234a8d1404af42e8cc20/packages/stylelint-plugin/src/rules/no-prefix-literal.ts#L42
- https://github.com/palantir/blueprint/blob/67c9e2ec477f0c189782234a8d1404af42e8cc20/packages/stylelint-plugin/src/rules/no-color-literal.ts#L43

we should be ok making this change, especially since it's going to go on the 4.x `next` branch

**Comment by adidahiya** [comment]
Just realized this is targeting develop and instead it should target the "next" branch, since these are breaking changes and need to be released with 4.0.

The code changes look good at this point, so in order to get this merged here are the next steps:

- [x] I will merge `develop` into `next` (I can do this some time this week)
- [x] Then we'll need to merge `next` into this branch
- [x] Finally, re-target this PR to `next`

**Comment by ianpf** [comment]
Sorry, realized that my sass-compile script didn't handle single file watching properly. Just wanted to sort that out.

**Comment by adidahiya** [comment]
Sorry for the build churn here. Regenerating the lockfile when I switched branches to `next` made the node modules post-install fail because of an old Python version in the CI container. Should be fixed now since I bumped the CI image tag. Working on fixing the rest of the build now.

**Comment by adidahiya** [comment]
Most things are working now, except for the unit test suite. It seems like chrome headless in the new CI image isn't happy running our tests. I'll look into this and try to extract all these unrelated changes out from this PR...

- [ ] fix chrome headless timeouts
- [ ] extract CI tooling changes to a separate PR

**Comment by til-schneider** [comment]
> I think that I can probably use the old svg-inline package, if I change to used the JavaScript API for the `Sass` package. It looks like it supports node-sass like plugins here: https://sass-lang.com/documentation/js-api#functions

I don't know whether you still struggle with inline SVGs. Here is my solution to this problem for my custom-blueprint build:

```ts
import sass from 'sass'

[...]

const sassOptions: sass.Options = {
    functions: {
        /**
         * Sass function to inline a UI icon svg and change its path color.
         *
         * Usage:
         * svg-icon("16px/icon-name.svg", (path: (fill: $color)) )
         */
        'svg-icon($path, $selectors: null)': function(path: sass.types.SassType,
            selectors: sass.types.SassType): sass.types.SassType
        {
            var pathJson = sassTypeToJson(path)
            var selectorsJson = sassTypeToJson(selectors)
            var encodedSelectors = JSON.stringify(selectorsJson)
            //console.log(`## svg-icon - pathJson: ${JSON.stringify(pathJson)}, selectorsJson: ${encodedSelectors}`)
            try {
                // Values are taken from node_modules/@blueprintjs/core/lib/css/blueprint.css
                var resultStr: string
                if (pathJson === '16px/chevron-right.svg' && encodedSelectors === '{"path":{"fill":"#5c7080"}}') {
                    resultStr = `url("data:image/svg+xml,%3csvg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 16 16'%3e%3cpath fill-rule='evenodd' clip-rule='evenodd' d='M10.71 7.29l-4-4a1.003 1.003 0 00-1.42 1.42L8.59 8 5.3 11.29c-.19.18-.3.43-.3.71a1.003 1.003 0 001.71.71l4-4c.18-.18.29-.43.29-.71 0-.28-.11-.53-.29-.71z' fill='%235C7080'/%3e%3c/svg%3e")`
                } else if (pathJson === '16px/more.svg' && encodedSelectors === '{"circle":{"fill":"#5c7080"}}') {
                    resultStr = `url("data:image/svg+xml,%3csvg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 16 16'%3e%3cg fill='%235C7080'%3e%3ccircle cx='2' cy='8.03' r='2'/%3e%3ccircle cx='14' cy='8.03' r='2'/%3e%3ccircle cx='8' cy='8.03' r='2'/%3e%3c/g%3e%3c/svg%3e")`
                } else if (pathJson === '16px/small-tick.svg' && encodedSelectors === '{"path":{"fill":"#ffffff"}}') {
                    resultStr = `url("data:image/svg+xml,%3csvg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 16 16'%3e%3cpath fill-rule='evenodd' clip-rule='evenodd' d='M12 5c-.28 0-.53.11-.71.29L7 9.59l-2.29-2.3a1.003 1.003 0 00-1.42 1.42l3 3c.18.18.43.29.71.29s.53-.11.71-.29l5-5A1.003 1.003 0 0012 5z' fill='white'/%3e%3c/svg%3e")`
                } else if (pathJson === '16px/small-minus.svg' && encodedSelectors === '{"path":{"fill":"#ffffff"}}') {
                    resultStr = `url("data:image/svg+xml,%3csvg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 16 16'%3e%3cpath fill-rule='evenodd' clip-rule='evenodd' d='M11 7H5c-.55 0-1 .45-1 1s.45 1 1 1h6c.55 0 1-.45 1-1s-.45-1-1-1z' fill='white'/%3e%3c/svg%3e")`
                } else {
                    throw new Error(`Unknown svg-icon: path='${JSON.stringify(pathJson)}', selectors='${encodedSelectors}'`)
                }
                return new sass.types.String(resultStr)
            } catch (error) {
                throw new Error(`Could not load sass svg-icon: ${JSON.stringify(pathJson)} (selectors: ${encodedSelectors}) - ${error}`)
            }
        },
    },
}


function sassTypeToJson(value: sass.types.SassType): any {
    if (value instanceof sass.types.Number) {
        return `${value.getValue()}${value.getUnit()}`
    } else if (value instanceof sass.types.String) {
        return value.getValue()
    } else if (value instanceof sass.types.Map) {
        var result: any = {}
        for (var i = 0, il = value.getLength(); i < il; i++) {
            result[sassTypeToJson(value.getKey(i))] = sassTypeToJson(value.getValue(i))
        }
        return result
    } else if (value instanceof sass.types.Color) {
        if (value.getA() !== 1) {
            throw new Error(`node-sass colors with transparence not supported`)
        }
        var r = value.getR().toString(16)
        var g = value.getG().toString(16)
        var b = value.getB().toString(16)

        function fill(value: string): string {
            return value.length === 1 ? `0${value}` : value
        }

        return `#${fill(r)}${fill(g)}${fill(b)}`
    } else {
        throw new Error(`Unsupported node-sass value type: ${typeof value.constructor.name}`)
    }
}
```

**Comment by ianpf** [comment]
Hey,

Sorry I let this linger so long. I got everything up to date with the v5 branch so that merging should be possible. That being said there's a failing test I can't replicate locally.

Let me know if there's anything else you need me to do (including possibly cleaning up the commit history, eep sorry).

**Comment by ianpf** [comment]
@adidahiya 
Hey, just wondering if there's any potential for this to be included in v5. I know there's currently a failing test, but I can't reproduce the failure locally, and I don't have permissions to rerun the job.

Thanks,

**Comment by adidahiya** [comment]
@ianpf thanks for revisiting and updating this PR. I'm definitely tracking it for v5.0, I've just been busy finishing up v4.0 things this month. I want to get this merged soon, likely by end of February. After that, I'm aiming for v5.0 to release in ~April/May.

File: packages/core/src/components/menu/_submenu.scss [hunk 9]
```diff
@@ -16,7 +16,7 @@
     &.#{$ns}-popover-open > .#{$ns}-menu-item {
       // keep a trail of hovered items as submenus are opened
       /* stylelint-disable-next-line scss/at-extend-no-missing-placeholder */
-      @extend .#{$ns}-menu-item:hover;
+      @extend .#{$ns}-menu-item, :hover;
```
Comment (lwouis) [inline packages/core/src/components/menu/_submenu.scss]: Have you tested this end-to-end (i.e. visually by playing around `Menu` with nested `MenuItem`)? [More details here](https://github.com/palantir/blueprint/issues/4032#issuecomment-1039940037)

Comment (ianpf) [reply packages/core/src/components/menu/_submenu.scss]: Thanks for finding this.
We can't just not make this change, as compound selector extends are not valid sass in dart-sass.

I've used a placeholder value, as suggested [here](https://sass-lang.com/documentation/breaking-changes/extend-compound) in the "Heads Up!" section to get around the issue, but that might have it's own issues.

**Comment by chriswiggins** [comment]
FYI I've been working on the @vgrid/sass-inline-svg package (branch deps-update), which uses the latest "modern" api to compile sass spreadsheets. I've now halted my progress as I've seen this PR, and trying to work around all of the issues in webpack. The tests work as expected in the inliner module, but I haven't pushed it to npm yet. Perhaps you'd like to try this if you're wanting to move to the modern `sass` api?

I might even look into making the module compatible with both modern and legacy apis

**Comment by chriswiggins** [comment]
Theoretically, the https://github.com/vgrid/sass-inline-svg module should now support the legacy api too, and is currently on npm with version `2.0.0-0` and tag `next`. Let me know if you have a chance to test and whether or not this is working as expected.

**Comment by adidahiya** [comment]
Hey @ianpf thanks for all your work over this over the months while this PR has been open. I decided to include the migration to dart-sass in v4.x instead of v5.x (see https://github.com/palantir/blueprint/issues/4032#issuecomment-1086175879). Apologies that the roadmap has shifted and we allowed this PR to get so out of sync with the repo. We have an alternative implementation in #5216, which I've now merged. Please try it out with the next minor release and let me know if you run into any issues.
